### DISCUSSION TITLE
How to use `__class_getitem__`?

### DISCUSSION THREAD

**Discussion by A-kirami** [microsoft/pylance-release#4478] [Q&A]
I'm trying to use it like the following.

```python
from typing import TypeVar

T = TypeVar("T")


class Ref:
    def __class_getitem__(cls, item: T) -> T:
        return item


a: Ref[int] = 0
```

But there was an error reported by pylance:

    Expression of type "Literal[0]" cannot be assigned to declared type "Ref"
      "Literal[0]" is incompatible with "Ref"

I want it to be like `typing. Annotated`, how should I modify it?

**Comment by erictraut** [comment]
First, you're using `__class_getitem__` in a manner for which it wasn't designed. Please refer to [this documentation](https://docs.python.org/3/reference/datamodel.html#the-purpose-of-class-getitem) which specifically warns against this.

Second, you're using the expression `Ref[int]` in a type annotation for a variable `a`. The annotation `Ref[int]` is going to be interpreted by a type checker as "an instance of `Ref` parameterized with a type argument `int`. That means `a` must be assigned an instance of `Ref`, but you're attempting to assign it an instance of `int`, which a type checker will prevent.

Could you provide more information about what you're trying to do? If you want `Ref` to be generic, you should use the standard `Generic` base class, like this:
```python
class Ref(Generic[T]): ...
```
